### DISCUSSION TITLE
The official game demo, named `MemoryTile`, the CPU usage is very high (around 30 percent).

### DISCUSSION THREAD

**Discussion by bowenxuuu** [slint-ui/slint#5677] [Q&A] [answered]
I have developed a game demo named `MemoryTile`, following the official website's documentation. However, both in `release` and `debug` build modes, the CPU usage is quite high (around 30 percent).
﻿
I suspect that my program is not utilizing a GPU rendering backend, but instead is using CPU software rendering, which is why the CPU usage is so high. Could you please give me some advice on what steps I should take to ensure that Slint uses GPU rendering backends like `skia` or `femtovg` during runtime?

Lastly, I would like to know if `Slint` currently supports using `wgpu` as a GPU rendering backend. As far as I know, some popular GUI frameworks such as `iced` and `egui` already support `wgpu`. If Slint also supports it, could you guide me on how to configure it as the renderer backend?

![微信图片_20240723172058](https://github.com/user-attachments/assets/3cdb8217-2f2d-45e2-8d2b-2528737e9db4)

```
// Cargo.toml

[package]
name = "slint_learn"
version = "0.1.0"
edition = "2021"
publish = false
build = "build.rs"

[dependencies]
slint = "1.7"
rand = "0.8"

[build-dependencies]
slint-build = "1.7"

[profile.release]
lto = true
opt-level = 3
panic = "abort"
codegen-units = 1
```

```
// appwindow.slint

import { Button, AboutSlint } from "std-widgets.slint";

struct TileData {
    image: image,
    image_visible: bool,
    solved: bool,
}

component MemoryTile inherits Rectangle {
    callback clicked;
    in property <bool> open_curtain;
    in property <bool> solved;
    in property <image> icon;
    height: 64px;
    width: 64px;
    border-radius: 6px;
    background: solved ? #34CE57 : #3960D5;
    animate background { duration: 800ms; }
    Image {
        source: icon;
        width: parent.width;
        height: parent.height;
    }

    // Left curtain
    Rectangle {
        background: #193076;
        x: 0px;
        width: open_curtain ? 0px : (parent.width / 2);
        height: parent.height;
        border-top-left-radius: 6px;
        border-bottom-left-radius: 6px;
        animate width {
            duration: 250ms;
            easing: ease-in;
        }
    }

    // Right curtain
    Rectangle {
        background: #193076;
        x: open_curtain ? parent.width : (parent.width / 2);
        width: open_curtain ? 0px : (parent.width / 2);
        height: parent.height;
        border-top-right-radius: 6px;
        border-bottom-right-radius: 6px;
        animate x, width {
            duration: 250ms;
            easing: ease-in;
        }
    }

    TouchArea {
        clicked => {
            root.clicked();
        }
    }
}

component MainWindow inherits Window {
    width: 306px;
    height: 306px;
    callback check_if_pair_solved();
    in property <bool> disable_tiles;
    in property <[TileData]> memory_tiles: [
        { image: @image-url("../icons/at.png") },
        { image: @image-url("../icons/balance-scale.png") },
        { image: @image-url("../icons/bicycle.png") },
        { image: @image-url("../icons/bus.png") },
        { image: @image-url("../icons/cloud.png") },
        { image: @image-url("../icons/cogs.png") },
        { image: @image-url("../icons/motorcycle.png") },
        { image: @image-url("../icons/video.png") },
    ];
    for tile[i] in memory_tiles: MemoryTile {
        x: mod(i, 4) * 74px + 10px;
        y: floor(i / 4) * 74px + 10px;
        width: 64px;
        height: 64px;
        icon: tile.image;
        open_curtain: tile.image_visible || tile.solved;
        // propagate the solved status from the model to the tile
        solved: tile.solved;
        clicked => {
            // tile.image_visible = !tile.image_visible;
            if (!root.disable_tiles) {
                tile.image_visible = !tile.image_visible;
                root.check_if_pair_solved();
            }
        }
    }
}

export { MainWindow }
```

```
// main.rs

#![cfg_attr(not(debug_assertions), windows_subsystem = "windows")] // hide console window on Windows in release

use rand::seq::SliceRandom;
use slint::Model;

slint::include_modules!();

fn main() {
    let main_window = MainWindow::new().unwrap();

    // Fetch the tiles from the model
    let mut tiles: Vec<TileData> = main_window.get_memory_tiles().iter().collect();
    // Duplicate them to ensure that we have pairs
    tiles.extend(tiles.clone());

    // Randomly mix the tiles

    let mut rng = rand::thread_rng();
    tiles.shuffle(&mut rng);

    // Assign the shuffled Vec to the model property
    let tiles_model = std::rc::Rc::new(slint::VecModel::from(tiles));
    main_window.set_memory_tiles(tiles_model.clone().into());

    let main_window_weak = main_window.as_weak();
    main_window.on_check_if_pair_solved(move || {
        let mut flipped_tiles = tiles_model
            .iter()
            .enumerate()
            .filter(|(_, tile)| tile.image_visible && !tile.solved);

        if let (Some((t1_idx, mut t1)), Some((t2_idx, mut t2))) =
            (flipped_tiles.next(), flipped_tiles.next())
        {
            let is_pair_solved = t1 == t2;
            if is_pair_solved {
                t1.solved = true;
                tiles_model.set_row_data(t1_idx, t1);
                t2.solved = true;
                tiles_model.set_row_data(t2_idx, t2);
            } else {
                let main_window = main_window_weak.unwrap();
                main_window.set_disable_tiles(true);
                let tiles_model = tiles_model.clone();
                slint::Timer::single_shot(std::time::Duration::from_secs(1), move || {
                    main_window.set_disable_tiles(false);
                    t1.image_visible = false;
                    tiles_model.set_row_data(t1_idx, t1);
                    t2.image_visible = false;
                    tiles_model.set_row_data(t2_idx, t2);
                });
            }
        }
    });

    main_window.run().unwrap();
}
```

```
// build.rs

fn main() {
    slint_build::compile("ui/appwindow.slint").unwrap();
}
```

**Comment by tronical** [comment, marked as answer]
> I have developed a game demo named `MemoryTile`, following the official website's documentation. However, both in `release` and `debug` build modes, the CPU usage is quite high (around 30 percent). ﻿ I suspect that my program is not utilizing a GPU rendering backend, but instead is using CPU software rendering, which is why the CPU usage is so high. 

We default to rendering with FemtoVG on Windows, which uses the OpenGL drivers that you have installed on your system. If that weren't available, you'd see a message in the console that indicates a fallback to software rendering. Do you see that?

> Could you please give me some advice on what steps I should take to ensure that Slint uses GPU rendering backends like `skia` or `femtovg` during runtime?

Adding `renderer-skia` to the features in the slint crate (`slint = { version = "1.7", features = ["renderer-skia"] }`) will automatically prefer Skia over FemtoVG.

> Lastly, I would like to know if `Slint` currently supports using `wgpu` as a GPU rendering backend. As far as I know, some popular GUI frameworks such as `iced` and `egui` already support `wgpu`. If Slint also supports it, could you guide me on how to configure it as the renderer backend?

FemtoVG does not support `wgpu` yet, and Skia uses internally an architecture that's identical to `wgpu` - it abstracts over different underlying graphics APIs such as Metal, Direct3D, Vulkan, and OpenGL.

I suggest to run the program in a profiler to get an idea where the time might be spent.

**Comment by bowenxuuu** [comment]
> We default to rendering with FemtoVG on Windows, which uses the OpenGL drivers that you have installed on your system. If that weren't available, you'd see a message in the console that indicates a fallback to software rendering. Do you see that?

No, I don't see that message. I've checked the console output thoroughly and there's no indication of a fallback to software rendering.

> Adding renderer-skia to the features in the slint crate (slint = { version = "1.7", features = ["renderer-skia"] }) will automatically prefer Skia over FemtoVG.

After adding the `renderer-skia`, the CPU usage has notably decreased, dropping from approximately `30 percent` to the current rate of around `2 percent`. However, the memory consumption has increased, now using around `60MB`, whereas it was previously using about `20MB`. The size of the compiled binary executable has also increased, now being roughly `20MB`, compared to about `3MB` before. These findings were noted during tests executed in `release` mode.

**Reply by tronical** [reply]
That sounds like it's in the expected ballpark. What kind of GPU and drivers do you have, btw?

**Reply by bowenxuuu** [reply]
> That sounds like it's in the expected ballpark. What kind of GPU and drivers do you have, btw?

```
System: Windows 10
CPU: Intel i7-8557U
GPU: Intel(R) Iris(R) Plus Graphics 645
DirectX Version: 12 (FL 12.1)
Driver Version: 31.0.101.2111
```

Slint is a great GUI framework for me, though I do have some minor suggestions:

- The mutual invocation between Slint code and Rust code isn’t very convenient, or even somewhat difficult. (This might be due to my limited knowledge, as I’ve only been learning Slint for a short time.)
- It would be great to provide multi-language documentation, which could help Slint become more popular across different countries, as not everyone’s native language is English.
- Continuously optimize performance and increase maturity, which I believe should be a long-term goal for Slint.

Finally, I wish Slint the best in its development!

**Reply by tronical** [reply]
Thank you for your words of encouragement :).

> The mutual invocation between Slint code and Rust code isn’t very convenient, or even somewhat difficult. (This might be due to my limited knowledge, as I’ve only been learning Slint for a short time.)

Agreed, this is an area that needs improvement and we're working on it.

> It would be great to provide multi-language documentation, which could help Slint become more popular across different countries, as not everyone’s native language is English.

This comes up occasionally, but I'm unsure how to deal with it, to be honest. On the one hand, the documentation and product changes at a fast pace still, so translations will be outdated sooner than later. On the other hand, you have a point in that English is not everyone's native language, and it would be good to reach a wide audience.

Do you know how other projects handle this? Maybe there's something that we can learn?

Do you think that there might be any value in a completely machine-generated translation (on-the-fly perhaps)?

**Reply by bowenxuuu** [reply]
Translating multi-language documents is indeed a tedious task. Perhaps we could try machine translation followed by human proofreading, and also invite community members to help with the translation. Beyond that, I don't have any better suggestions.

**Reply by tronical** [reply]
Right, maybe a translated version could be in three states:

1. Initially machine translated. A banner could indicate that this hasn't been proofread yet.
2. Human reviewed. No banner required, this is all good.
3. It might be (1) or (2), but there's a newer version available in the English original. A banner could indicate that the documentation might be outdated.

**Reply by bowenxuuu** [reply]
Yes, during the first state, a link to the English content can be placed next to the translated content, making it convenient for people to compare the original content.
